### PR TITLE
KEP-4222: Fall back to JSON request encoding after CBOR 415.

### PR DESCRIPTION
**Pull request by benluddy**
#### What type of PR is this?

/kind cleanup
/sig api-machinery

#### What this PR does / why we need it:

If a client is configured to encode request bodies to CBOR, but the server does not support CBOR, the server will respond with HTTP 415 (Unsupported Media Type). By feeding this response back to the RESTClient, subsequent requests can fall back to JSON, which is assumed to be acceptable.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://kep.k8s.io/4222
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/rest/client.go [hunk 1]
```diff
@@ -237,5 +237,40 @@ func (c *RESTClient) Delete() *Request {
 
 // APIVersion returns the APIVersion this RESTClient is expected to use.
 func (c *RESTClient) APIVersion() schema.GroupVersion {
-	return c.content.GroupVersion
+	return c.content.GetClientContentConfig().GroupVersion
+}
+
+type requestClientContentConfigProvider struct {
+	base    ClientContentConfig
+	tripped atomic.Bool
```
Comment (deads2k) [inline staging/src/k8s.io/client-go/rest/client.go:245]: give me a better name. If true, what is supposed to happen?  Is this `forceContentTypeToJSON`?

File: staging/src/k8s.io/client-go/rest/client.go [hunk 2]
```diff
@@ -237,5 +237,40 @@ func (c *RESTClient) Delete() *Request {
 
 // APIVersion returns the APIVersion this RESTClient is expected to use.
 func (c *RESTClient) APIVersion() schema.GroupVersion {
-	return c.content.GroupVersion
+	return c.content.GetClientContentConfig().GroupVersion
+}
+
+type requestClientContentConfigProvider struct {
```
[the remark this answers is not in the document]
Comment (deads2k) [reply staging/src/k8s.io/client-go/rest/client.go:243]: I think this is trying to:

```
// If the client requests CBOR and the server does not understand CBOR, force future requests to use
// JSON so they will succeed.  This is convenient for client users that want to prefer CBOR, but also
// need to interoperate with older servers so requests do not permanently fail.
// The clients will not default to using CBOR until at least all supported kube-apiservers have enable-CBOR
// locked to true, so this path will be rarely taken.
// Additionally, all generated clients accessing built-in kube resources are forced to protobuf, so those
// will not degrade to JSON.
```

[the remark this answers is not in the document]
Comment (benluddy) [reply staging/src/k8s.io/client-go/rest/client.go:250]: Thanks. That's all accurate. I incorporated your suggested language into a comment here.

File: staging/src/k8s.io/client-go/rest/client.go (hunk 1 above)
Comment (benluddy) [reply staging/src/k8s.io/client-go/rest/client.go]: Commented and renamed to `sawUnsupportedMediaTypeForCBOR` to describe how its value should be interpreted.

**Comment by deads2k** [comment]
Thanks, that helps clarify a lot and the code appears to do that.

/lgtm
/approve

**Pull request merged**